src/devices/RGBDSensorClient/RGBDSensorClient_StreamingMsgParser.cpp

```diff
@@ -9,67 +9,97 @@
 
 #include "RGBDSensorClient_StreamingMsgParser.h"
 
+#include <yarp/os/Time.h>
+
 using namespace yarp::dev;
 
 // Callback reader for rgb
-RgbImageReader_Impl::RgbImageReader_Impl() = default;
-RgbImageReader_Impl::~RgbImageReader_Impl() = default;
-
-void RgbImageReader_Impl::onRead(yarp::sig::FlexImage& datum)
+void RgbImageBufferedPort::onRead(yarp::sig::FlexImage& datum)
 {
-    last_rgb = datum;
+    local_arrival_time = yarp::os::Time::now();
+    std::lock_guard<std::mutex> lock(mutex);
+    std::swap(datum, last_rgb);
+    getEnvelope(stamp);
 }
 
-yarp::sig::FlexImage RgbImageReader_Impl::getImage()
+std::tuple<bool, yarp::sig::FlexImage, yarp::os::Stamp> RgbImageBufferedPort::getImage() const
 {
-    return last_rgb;
+    std::lock_guard<std::mutex> lock(mutex);
+    if (local_arrival_time <= 0.0) {
+        // No image received yet
+        return std::make_tuple(false, yarp::sig::FlexImage(), yarp::os::Stamp());
+    }
+    return std::make_tuple(true, last_rgb, stamp);
 }
 
 
 // callback reader for depthImage
-FloatImageReader_Impl::FloatImageReader_Impl() = default;
-FloatImageReader_Impl::~FloatImageReader_Impl() = default;
-
-void FloatImageReader_Impl::onRead(yarp::sig::ImageOf< yarp::sig::PixelFloat> & datum)
+void FloatImageBufferedPort::onRead(yarp::sig::ImageOf< yarp::sig::PixelFloat> & datum)
 {
-    last_depth = datum;
+    local_arrival_time = yarp::os::Time::now();
+    std::lock_guard<std::mutex> lock(mutex);
+    std::swap(datum, last_depth);
+    getEnvelope(stamp);
 }
 
-yarp::sig::ImageOf<yarp::sig::PixelFloat> FloatImageReader_Impl::getImage()
+std::tuple<bool, yarp::sig::ImageOf<yarp::sig::PixelFloat>, yarp::os::Stamp> FloatImageBufferedPort::getImage() const
 {
-    return last_depth;
+    std::lock_guard<std::mutex> lock(mutex);
+    if (local_arrival_time <= 0.0) {
+        // No image received yet
+        return std::make_tuple(false, yarp::sig::ImageOf<yarp::sig::PixelFloat>(), yarp::os::Stamp());
+    }
+    return std::make_tuple(true, last_depth, stamp);
```

Comment:
I believe here (and in other places like this) you should be allowed to reduce it to:

```c++
return {true, last_depth, stamp};
```

Reply:
You might want to disregard this comment... There is something fishy around `std::tuple`, look for "conditionally-explicit constructors" in https://en.cppreference.com/w/cpp/utility/tuple/tuple. According to https://stackoverflow.com/a/16713413, this was a compiler extension back in 2013 or so, but later (partially?) fixed in C++11 via [N4387](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4387.html).

Reply:
Yes, unfortunately N4387 is in C++17 and we still require c++14.
It's probably worth add a FIXME comment to remember changing it when we have the compiler support